From 2ee4804aec6843f00e77cf2083ae6ff01d6252fe Mon Sep 17 00:00:00 2001
From: Xaver Hellauer <xaver.hellauer@gmail.com>
Date: Sun, 23 Aug 2020 23:11:38 +0200
Subject: [PATCH] Icon Theme

---
 src/icons.h | 381 +++++++++++++++++++++++++++++-----------------------
 src/nnn.c   |   8 +-
 2 files changed, 217 insertions(+), 172 deletions(-)

diff --git a/src/icons.h b/src/icons.h
index 255ad37..79f82d3 100644
--- a/src/icons.h
+++ b/src/icons.h
@@ -1,5 +1,3 @@
-#include "icons-in-terminal.h"
-
 struct icon_pair {
 	const char *match;
 	const char *icon;
@@ -15,14 +13,15 @@ struct icon_pair {
  * Adjust if the icons are not printed properly
  */
 
-#define ICON_PADDING_LEFT  ""
-#define ICON_PADDING_RIGHT " "
+#define ICON_PADDING_LEFT  " "
+#define ICON_PADDING_RIGHT "  "
 
-#define VIDEO_COLOR    93 /* Purple */
-#define AUDIO_COLOR   220 /* Gold1 */
-#define IMAGE_COLOR    82 /* Chartreuse2 */
-#define BOOKS_COLOR   202 /* OrangeRed1 */
-#define ARCHIVE_COLOR 209 /* Salmon1 */
+#define VIDEO_COLOR       0x44
+#define AUDIO_COLOR       0x46
+#define IMAGE_COLOR       0x2c
+#define BOOKS_COLOR       0x21
+#define ARCHIVE_COLOR     12
+#define UNIMPORTANT_COLOR 8
 
 /*
  * Using symbols defined in icons-in-terminal.h, or even using icons-in-terminal is not necessary.
@@ -30,26 +29,35 @@ struct icon_pair {
  * If you are using icons-in-terminal the creator recommends that you do use the symbols in the generated header.
  */
 
-static const struct icon_pair dir_icon  = {"", FA_FOLDER, 0};
-static const struct icon_pair file_icon = {"", FA_FILE_O, 0};
-static const struct icon_pair exec_icon = {"", FA_COG,    0};
+static const struct icon_pair dir_icon = {"", "", 0};
+static const struct icon_pair file_icon = {"", "", 0};
+static const struct icon_pair exec_icon = {"", "", 0};
 
 /* All entries are case-insensitive */
 
 static const struct icon_pair icons_name[] = {
-	{".git",         FA_GIT,        0},
-	{"Desktop",      FA_DESKTOP,    0},
-	{"Documents",    FA_BRIEFCASE,  0},
-	{"Downloads",    FA_DOWNLOAD,   0},
-	{"Music",        FA_MUSIC,      0},
-	{"Pictures",     MD_CAMERA_ALT, 0},
-	{"Public",       FA_INBOX,      0},
-	{"Templates",    FA_PAPERCLIP,  0},
-	{"Videos",       FA_FILM,       0},
-	{"CHANGELOG",    FA_HISTORY,    0},
-	{"configure",    FILE_CONFIG,   0},
-	{"License",      FA_COPYRIGHT,  0},
-	{"Makefile",     FILE_CMAKE,    0},
+	{".git",             "",          0},
+	{"docs",             "",          0},
+	{"node_modules",     "",          0},
+	{"Music",            "",          0},
+	{"pix",              "",          0},
+	{"src",              "",          0},
+	{"mov",              "",          0},
+	{"CHANGELOG",        "",          0},
+	{"config",           "",          0},
+	{"configure",        "",          0},
+	{"CMakeLists.txt",   "",          0},
+	{"License",          "",          UNIMPORTANT_COLOR},
+	{"Makefile",         "",          0},
+	{"gulpfile.js",      "",          0},
+	{"Gruntfile.js",     "",          0},
+	{"Gruntfile.coffee", "",          0},
+	{"Dockerfile",       "",          0},
+	{".vscode",          "",          0},
+	{"nix",              "",          0},
+	{".npmrc",           "",          0},
+	{"package-lock.json","",          0},
+	{"package.json",     "",          UNIMPORTANT_COLOR},
 };
 
 /*
@@ -60,201 +68,238 @@ static const struct icon_pair icons_name[] = {
 
 static const struct icon_pair icons_ext[] = {
 	/* Numbers */
-	{"1",        FILE_MANPAGE,         0},
-	{"7z",       FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
+	{"1",        "",                 0},
+	{"2",        "",                 0},
+	{"3",        "",                 0},
+	{"4",        "",                 0},
+	{"5",        "",                 0},
+	{"6",        "",                 0},
+	{"7",        "",                 0},
+	{"7z",       "",                 ARCHIVE_COLOR},
 
 	/* A */
-	{"a",        FILE_MANPAGE,         0},
-	{"apk",      FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
-	{"asm",      FILE_NASM,            0},
-	{"aup",      FA_FILE_AUDIO_O,      AUDIO_COLOR},
-	{"avi",      FA_FILE_MOVIE_O,      VIDEO_COLOR},
+	{"a",        "",                 0},
+	{"ai",       "",                 IMAGE_COLOR},
+	{"apk",      "",                 ARCHIVE_COLOR},
+	{"aup",      "",                 AUDIO_COLOR},
+	{"avi",      "",                 VIDEO_COLOR},
+	{"azw",      "",                 BOOKS_COLOR},
+	{"azw3",     "",                 BOOKS_COLOR},
 
 	/* B */
-	{"bat",      MFIZZ_SCRIPT,         0},
-	{"bin",      OCT_FILE_BINARY,      0},
-	{"bmp",      FA_FILE_IMAGE_O,      IMAGE_COLOR},
-	{"bz2",      FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
+	{"bat",      "",                 0},
+	{"bmp",      "",                 IMAGE_COLOR},
+	{"bz2",      "",                 ARCHIVE_COLOR},
 
 	 /* C */
-	{"c",        MFIZZ_C,              0},
-	{"c++",      MFIZZ_CPLUSPLUS,      0},
-	{"cab",      FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
-	{"cbr",      FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
-	{"cbz",      FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
-	{"cc",       MFIZZ_CPLUSPLUS,      0},
-	{"class",    MFIZZ_JAVA,           0},
-	{"clj",      MFIZZ_CLOJURE,        0},
-	{"cljc",     MFIZZ_CLOJURE,        0},
-	{"cljs",     MFIZZ_CLOJURE,        0},
-	{"cmake",    FILE_CMAKE,           0},
-	{"coffee",   MFIZZ_COFFEE_BEAN,    0},
-	{"conf",     FA_COGS,              0},
-	{"cpio",     FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
-	{"cpp",      MFIZZ_CPLUSPLUS,      0},
-	{"css",      MFIZZ_CSS3,           0},
-	{"cue",      FA_FILE_AUDIO_O,      AUDIO_COLOR},
-	{"cvs",      FA_COGS,              0},
-	{"cxx",      MFIZZ_CPLUSPLUS,      0},
+	{"c",        "",                 0},
+	{"c++",      "",                 0},
+	{"cab",      "",                 ARCHIVE_COLOR},
+	{"cb7",      "",                 BOOKS_COLOR},
+	{"cbr",      "",                 BOOKS_COLOR},
+	{"cbz",      "",                 BOOKS_COLOR},
+	{"cc",       "",                 0},
+	{"class",    "",                 0},
+	{"clj",      "",                 0},
+	{"cljc",     "",                 0},
+	{"cljs",     "",                 0},
+	{"cmake",    "",                 0},
+	{"coffee",   "",                 0},
+	{"conf",     "",                 0},
+	{"cpio",     "",                 ARCHIVE_COLOR},
+	{"cpp",      "",                 0},
+	{"cs",       "",                 0},
+	{"css",      "",                 0},
+	{"cvs",      "",                 0},
+	{"cxx",      "",                 0},
 
 	/* D */
-	{"db",       MFIZZ_DATABASE_ALT2,  0},
-	{"deb",      MFIZZ_DEBIAN,         ARCHIVE_COLOR},
-	{"diff",     FILE_DIFF,            0},
-	{"dll",      FILE_MANPAGE,         0},
-	{"doc",      FILE_WORD,            0},
-	{"docx",     FILE_WORD,            0},
+	{"d",        "",                 0},
+	{"dart",     "",                 0},
+	{"db",       "",                   0},
+	{"deb",      "",                 ARCHIVE_COLOR},
+	{"diff",     "",                 0},
+	{"dll",      "",                 0},
+	{"dmg",      "",                 ARCHIVE_COLOR},
+	{"doc",      "",                 0},
+	{"docx",     "",                 0},
 
 	 /* E */
-	{"ejs",      FA_FILE_CODE_O,       0},
-	{"elf",      FA_LINUX,             0},
-	{"epub",     FA_FILE_PDF_O,        BOOKS_COLOR},
-	{"exe",      FA_WINDOWS,           0},
+	{"ejs",      "",                 0},
+	{"elf",      "",                 0},
+	{"epub",     "",                 BOOKS_COLOR},
+	{"erl",      "",                 0},
+	{"exe",      "",                 0},
 
 	/* F */
-	{"f#",       DEV_FSHARP,           0},
-	{"flac",     FA_FILE_AUDIO_O,      AUDIO_COLOR},
-	{"flv",      FA_FILE_MOVIE_O,      VIDEO_COLOR},
-	{"fs",       DEV_FSHARP,           0},
-	{"fsi",      DEV_FSHARP,           0},
-	{"fsscript", DEV_FSHARP,           0},
-	{"fsx",      DEV_FSHARP,           0},
+	{"f#",       "",                 0},
+	{"flac",     "",                 AUDIO_COLOR},
+	{"flv",      "",                 VIDEO_COLOR},
+	{"fs",       "",                 0},
+	{"fsi",      "",                 0},
+	{"fsscript", "",                 0},
+	{"fsx",      "",                 0},
 
 	/* G */
-	{"gem",      FA_FILE_ARCHIVE_O,    0},
-	{"gif",      FA_FILE_IMAGE_O,      IMAGE_COLOR},
-	{"go",       MFIZZ_GO,             0},
-	{"gz",       FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
-	{"gzip",     FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
+	{"gem",      "",                 0},
+	{"gif",      "",                 IMAGE_COLOR},
+	{"go",       "",                 0},
+	{"gz",       "",                 ARCHIVE_COLOR},
+	{"gzip",     "",                 ARCHIVE_COLOR},
 
 	/* H */
-	{"h",        MFIZZ_C,              0},
-	{"hh",       MFIZZ_CPLUSPLUS,      0},
-	{"htaccess", FA_COGS,              0},
-	{"htpasswd", FA_COGS,              0},
-	{"htm",      FA_FILE_CODE_O,       0},
-	{"html",     FA_FILE_CODE_O,       0},
-	{"hxx",      MFIZZ_CPLUSPLUS,      0},
+	{"h",        "",                 0},
+	{"hh",       "",                 0},
+	{"hrl",      "",                 0},
+	{"hs",       "",                 0},
+	{"htaccess", "",                 0},
+	{"htpasswd", "",                 0},
+	{"htm",      "",                 0},
+	{"html",     "",                 0},
+	{"hxx",      "",                 0},
 
 	/* I */
-	{"ico",      FA_FILE_IMAGE_O,      IMAGE_COLOR},
-	{"img",      FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
-	{"ini",      FA_COGS,              0},
-	{"iso",      LINEA_MUSIC_CD,       ARCHIVE_COLOR},
+	{"ico",      "",                 IMAGE_COLOR},
+	{"img",      "",                 ARCHIVE_COLOR},
+	{"ini",      "",                 0},
+	{"iso",      "",                 ARCHIVE_COLOR},
 
 	/* J */
-	{"jar",      MFIZZ_JAVA,           0},
-	{"java",     MFIZZ_JAVA,           0},
-	{"jl",       FA_COGS,              0},
-	{"jpeg",     FA_FILE_IMAGE_O,      IMAGE_COLOR},
-	{"jpg",      FA_FILE_IMAGE_O,      IMAGE_COLOR},
-	{"js",       DEV_JAVASCRIPT_BADGE, 0},
-	{"json",     MFIZZ_JAVASCRIPT,     0},
-	{"jsx",      FILE_JSX,             0},
+	{"jar",      "",                 0},
+	{"java",     "",                 0},
+	{"jl",       "",                 0},
+	{"jpeg",     "",                 IMAGE_COLOR},
+	{"jpg",      "",                 IMAGE_COLOR},
+	{"js",       "",                 0},
+	{"json",     "",                 0},
+	{"jsx",      "",                 0},
 
 	/* K */
 
 	/* L */
-	{"lha",      FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
-	{"log",      FA_FILE_TEXT_O,       0},
-	{"lua",      FILE_LUA,             0},
-	{"lzh",      FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
-	{"lzma",     FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
+	{"less",     "",                 0},
+	{"lha",      "",                 ARCHIVE_COLOR},
+	{"log",      "",                 0},
+	{"lua",      "",                 0},
+	{"lzh",      "",                 ARCHIVE_COLOR},
+	{"lzma",     "",                 ARCHIVE_COLOR},
 
 	/* M */
-	{"m4a",      FA_FILE_AUDIO_O,      AUDIO_COLOR},
-	{"m4v",      FA_FILE_MOVIE_O,      VIDEO_COLOR},
-	{"markdown", DEV_MARKDOWN,         0},
-	{"md",       DEV_MARKDOWN,         0},
-	{"mk",       FILE_CMAKE,           0},
-	{"mkv",      FA_FILE_MOVIE_O,      VIDEO_COLOR},
-	{"mov",      FA_FILE_MOVIE_O,      VIDEO_COLOR},
-	{"mp3",      FA_FILE_AUDIO_O,      AUDIO_COLOR},
-	{"mp4",      FA_FILE_MOVIE_O,      VIDEO_COLOR},
-	{"mpeg",     FA_FILE_MOVIE_O,      VIDEO_COLOR},
-	{"mpg",      FA_FILE_MOVIE_O,      VIDEO_COLOR},
-	{"msi",      FA_WINDOWS,           0},
+	{"m4a",      "",                 AUDIO_COLOR},
+	{"m4v",      "",                 VIDEO_COLOR},
+	{"markdown", "",                 0},
+	{"md",       "",                 0},
+	{"mk",       "",                 0},
+	{"mkv",      "",                 VIDEO_COLOR},
+	{"mobi",     "",                 BOOKS_COLOR},
+	{"mov",      "",                 VIDEO_COLOR},
+	{"mp3",      "",                 AUDIO_COLOR},
+	{"mp4",      "",                 VIDEO_COLOR},
+	{"mpeg",     "",                 VIDEO_COLOR},
+	{"mpg",      "",                 VIDEO_COLOR},
+	{"msi",      "",                 0},
 
 	/* N */
+	{"nar",       "",                 ARCHIVE_COLOR},
+	{"nix",       "",                0},
 
 	/* O */
-	{"o",          FILE_MANPAGE,       0},
-	{"ogg",        FA_FILE_AUDIO_O,    AUDIO_COLOR},
-	{"opdownload", FA_DOWNLOAD,        0},
-	{"out",        FA_LINUX,           0},
+	{"o",          "",               0},
+	{"oga",        "",               AUDIO_COLOR},
+	{"ogg",        "",               AUDIO_COLOR},
+	{"ogv",        "",               VIDEO_COLOR},
+	{"old",        "",               UNIMPORTANT_COLOR},
+	{"otf",        "",               0},
+	{"out",        "",               0},
 
 	/* P */
-	{"part",     FA_DOWNLOAD,          0},
-	{"patch",    FILE_PATCH,           0},
-	{"pdf",      FA_FILE_PDF_O,        BOOKS_COLOR},
-	{"php",      MFIZZ_PHP,            0},
-	{"png",      FA_FILE_IMAGE_O,      IMAGE_COLOR},
-	{"ppt",      FILE_POWERPOINT,      0},
-	{"pptx",     FILE_POWERPOINT,      0},
-	{"psb",      DEV_PHOTOSHOP,        0},
-	{"psd",      DEV_PHOTOSHOP,        0},
-	{"py",       MFIZZ_PYTHON,         0},
-	{"pyc",      MFIZZ_PYTHON,         0},
-	{"pyd",      MFIZZ_PYTHON,         0},
-	{"pyo",      MFIZZ_PYTHON,         0},
+	{"part",     "",                 UNIMPORTANT_COLOR},
+	{"patch",    "",                 0},
+	{"pax",      "",                 ARCHIVE_COLOR},
+	{"pdf",      "",                 BOOKS_COLOR},
+	{"php",      "",                 0},
+	{"pl",       "",                 0},
+	{"plx",      "",                 0},
+	{"pm",       "",                 0},
+	{"png",      "",                 IMAGE_COLOR},
+	{"ppt",      "",                 0},
+	{"pptx",     "",                 0},
+	{"ps",       "",                 BOOKS_COLOR},
+	{"psb",      "",                 0},
+	{"psd",      "",                 0},
+	{"py",       "",                 0},
+	{"pyc",      "",                 0},
+	{"pyd",      "",                 0},
+	{"pyo",      "",                 0},
 
 	/* Q */
 
 	/* R */
-	{"rar",      FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
-	{"rc",       FA_COGS,              0},
-	{"rom",      FA_LOCK,              0},
-	{"rpm",      FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
-	{"rss",      FA_RSS_SQUARE,        0},
-	{"rtf",      FA_FILE_PDF_O,        0},
+	{"rar",      "",                 ARCHIVE_COLOR},
+	{"rb",       "",                 0},
+	{"rc",       "",                 0},
+	{"rom",      "",                 0},
+	{"rpm",      "",                 ARCHIVE_COLOR},
+	{"rs",       "",                 0},
+	{"rss",      "",                 0},
+	{"rtf",      "",                 0},
 
 	/* S */
-	{"so",       FILE_MANPAGE,         0},
-	{"scala",    MFIZZ_SCALA,          0},
-	{"sh",       MFIZZ_SCRIPT,         0},
-	{"slim",     FA_FILE_CODE_O,       0},
-	{"sln",      DEV_VISUALSTUDIO,     0},
-	{"sql",      MFIZZ_MYSQL,          0},
-	{"srt",      FA_COMMENTS_O,        0},
-	{"sub",      FA_COMMENTS_O,        0},
-	{"svg",      FA_FILE_IMAGE_O,      IMAGE_COLOR},
+	{"sass",     "",                 0},
+	{"scss",     "",                 0},
+	{"scala",    "",                 0},
+	{"sh",       "",                 0},
+	{"sln",      "",                 0},
+	{"so",       "",                 0},
+	{"sql",      "",                 0},
+	{"srt",      "",                 0},
+	{"stylus",   "",                 0},
+	{"sub",      "",                 0},
+	{"svg",      "",                 IMAGE_COLOR}, // ﰟ
+	{"swift",    "",                 0},
 
 	/* T */
-	{"tar",      FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
-	{"tex",      FILE_TEX,             0},
-	{"tgz",      FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
-	{"ts",       FILE_TS,              0},
-	{"tsx",      FILE_TSX,             0},
-	{"txt",      FA_FILE_TEXT_O,       0},
-	{"txz",      FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
+	{"tar",      "",                 ARCHIVE_COLOR},
+	{"tex",      "",                 0},
+	{"tgz",      "",                 ARCHIVE_COLOR},
+	{"ts",       "",                 0},
+	{"tsx",      "",                 0},
+	{"ttc",      "",                 0},
+	{"ttf",      "",                 0},
+	{"txt",      "",                 0},
+	{"txz",      "",                 ARCHIVE_COLOR},
 
 	/* U */
 
 	/* V */
-	{"vid",      FA_FILE_MOVIE_O,      VIDEO_COLOR},
-	{"vim",      DEV_VIM,              0},
-	{"vimrc",    DEV_VIM,              0},
+	{"vid",      "",                 VIDEO_COLOR},
+	{"vim",      "",                 0},
+	{"vimrc",    "",                 0},
+	{"vue",      "",                 0}, // ﵂
 
 	/* W */
-	{"wav",      FA_FILE_AUDIO_O,      AUDIO_COLOR},
-	{"webm",     FA_FILE_MOVIE_O,      VIDEO_COLOR},
-	{"wma",      FA_FILE_AUDIO_O,      AUDIO_COLOR},
-	{"wmv",      FA_FILE_MOVIE_O,      VIDEO_COLOR},
+	{"wav",      "",                 AUDIO_COLOR},
+	{"webm",     "",                 VIDEO_COLOR},
+	{"wma",      "",                 AUDIO_COLOR},
+	{"wmv",      "",                 VIDEO_COLOR},
+	{"wv",       "",                 AUDIO_COLOR},
 
 	/* X */
-	{"xbps",     FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
-	{"xhtml",    FA_FILE_CODE_O,       0},
-	{"xls",      FILE_EXCEL,           0},
-	{"xlsx",     FILE_EXCEL,           0},
-	{"xml",      FA_FILE_CODE_O,       0},
-	{"xz",       FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR},
+	{"xbps",     "",                 ARCHIVE_COLOR},
+	{"xhtml",    "",                 0},
+	{"xls",      "",                 0},
+	{"xlsx",     "",                 0},
+	{"xml",      "",                 0},
+	{"xz",       "",                 ARCHIVE_COLOR},
 
 	/* Y */
-	{"yaml",     FA_COGS,              0},
-	{"yml",      FA_COGS,              0},
+	{"yaml",     "",                 0},
+	{"yml",      "",                 0},
 
 	/* Z */
-	{"zip",      FA_FILE_ARCHIVE_O,    ARCHIVE_COLOR}
+	{"Z",        "",                 ARCHIVE_COLOR},
+	{"zip",      "",                 ARCHIVE_COLOR},
+	{"zstd",     "",                 ARCHIVE_COLOR},
 
 	/* Other */
 };
diff --git a/src/nnn.c b/src/nnn.c
index be25850..7a6fb7e 100644
--- a/src/nnn.c
+++ b/src/nnn.c
@@ -1680,7 +1680,7 @@ static bool initcurses(void *oldmask)
 		use_default_colors();
 
 		/* Initialize file colors */
-		if (COLORS >= 256) {
+		if (COLORS >= 88) {
 			if (!(g_state.oldcolor || init_fcolors())) {
 				exitcurses();
 				fprintf(stderr, "NNN_FCOLORS!\n");
@@ -1695,7 +1695,7 @@ static bool initcurses(void *oldmask)
 		if (colors && *colors == '#') {
 			char *sep = strchr(colors, ';');
 
-			if (COLORS >= 256) {
+			if (COLORS >= 88) {
 				++colors;
 				ext = TRUE;
 
@@ -3524,17 +3524,17 @@ static void print_icon(const struct entry *ent, const int attrs)
 {
 	const struct icon_pair *picon = get_icon(ent);
 
-	addstr(ICON_PADDING_LEFT);
 	if (picon->color)
 		attron(COLOR_PAIR(C_UND + 1 + picon->color));
 	else if (attrs)
 		attron(attrs);
+	addstr(ICON_PADDING_LEFT);
 	addstr(picon->icon);
+	addstr(ICON_PADDING_RIGHT);
 	if (picon->color)
 		attroff(COLOR_PAIR(C_UND + 1 + picon->color));
 	else if (attrs)
 		attroff(attrs);
-	addstr(ICON_PADDING_RIGHT);
 }
 #endif
 
-- 
2.28.0

