diff -up a/dmenu.c b/dmenu.c
--- a/dmenu.c	2016-09-25 15:36:15.555089319 +0200
+++ b/dmenu.c	2016-09-25 15:39:46.169524704 +0200
@@ -595,6 +595,22 @@ setup(void)
 	                    DefaultVisual(dpy, screen),
 	                    CWOverrideRedirect | CWBackPixel | CWEventMask, &swa);

+	/* Set window properties */
+	if (win) {
+		{
+			char* NAME = "dmenu";
+			XTextProperty tprop_name;
+			if (Success == Xutf8TextListToTextProperty(dpy, &NAME, 1,
+			                                           XStringStyle,
+			                                           &tprop_name))
+				XSetWMName(dpy, win, &tprop_name);
+		}
+		{
+			XClassHint HINT = {"dmenu", "Dmenu"};
+			XSetClassHint(dpy, win, &HINT);
+		}
+	}
+
 	/* open input methods */
 	xim = XOpenIM(dpy, NULL, NULL, NULL);
 	xic = XCreateIC(xim, XNInputStyle, XIMPreeditNothing | XIMStatusNothing,
